### PR TITLE
adding lm eval script

### PR DESCRIPTION
**Pull request by bgoldberg-habana**
adding lm eval script

### CODE REVIEW

File: examples/text-generation/lm_eval.py [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+    args = setup_parser_eval()
+    model, tokenizer, generation_config = initialize_model(args, logger)
+
+    lm_tasks = lm_eval.tasks.get_task_dict(args.tasks)
+    lm = HabanaModelAdapter(tokenizer, model, args, generation_config)
+
+    eval_start = time.perf_counter()
+    results = lm_eval.evaluator.evaluate(lm, lm_tasks, limit=args.limit_iters)
+    if args.device == "hpu":
+        pass
```
Comment (polisettyvarma) [inline examples/text-generation/lm_eval.py:134]: what is the purpose ?

File: examples/text-generation/run_lm_eval.py [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+    def _model_call(self, inps):
+        seq_length = inps.shape[-1]
+        padding_length = 0
+        if self.options.static_shapes:
+            bucket_length = self.find_bucket(seq_length)
+            padding_length = bucket_length - seq_length
+            inps = F.pad(inps, (0, padding_length), value=self.model.config.pad_token_id)
+        logits = self.model(
+            inps.to(self._device),
+            use_cache=self.options.use_cache,
```
Comment (polisettyvarma) [inline examples/text-generation/run_lm_eval.py:129]: self.options.use_kv_cache

File: examples/text-generation/run_lm_eval.py [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+from utils import initialize_model
+
+from optimum.habana.utils import get_hpu_memory_stats
+
+
+os.environ.setdefault("TOKENIZERS_PARALLELISM", "false")
+logger = logging.getLogger(__name__)
+
+
+def setup_parser_eval():
```
Comment (polisettyvarma) [inline examples/text-generation/run_lm_eval.py:40]: setup_lm_eval_parser looks better to me, just a suggestion

File: examples/text-generation/run_lm_eval.py [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+            attn_softmax_bf16=self.options.attn_softmax_bf16,
+        )["logits"].cpu()
+        if self.options.static_shapes and padding_length > 0:
+            logits = logits[:, :-padding_length, :]
+        logits = logits.to(torch.float32)
+        return logits
+
+
+def main():
+    args = setup_parser_eval()
```
Comment (polisettyvarma) [inline examples/text-generation/run_lm_eval.py:139]: setup_lm_eval_parser looks better to me, just a suggestion

File: examples/text-generation/run_lm_eval.py [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+    lm_tasks = lm_eval.tasks.get_task_dict(args.tasks)
+    lm = HabanaModelAdapter(tokenizer, model, args, generation_config)
+
+    eval_start = time.perf_counter()
+    results = lm_eval.evaluator.evaluate(lm, lm_tasks, limit=args.limit_iters)
+    eval_end = time.perf_counter()
+
+    results["args"] = vars(args)
+    results["duration"] = eval_end - eval_start
+
```
Comment (polisettyvarma) [inline examples/text-generation/run_lm_eval.py:151]: it's good to have sync or barrier actually

**Review by polisettyvarma — CHANGES_REQUESTED** [review]

File: examples/text-generation/README.md [hunk 6]
```diff
@@ -212,3 +212,47 @@ python run_generation.py \
...
+Evaluate Llama 70B on 8 Gaudi2 on task WinoGrande, using the BF16 data type.
+```
+```
+deepspeed --num_gpus 8 run_lm_eval.py \
+--model_name_or_path meta-llama/Llama-2-70b-hf \
+--use_hpu_graphs \
+--use_kv_cache \
+--bf16 \
+--batch_size=1 \
+--task winogrande \
```
Comment (polisettyvarma) [inline examples/text-generation/README.md:256]: tasks

File: examples/text-generation/README.md [hunk 7]
```diff
@@ -212,3 +212,47 @@ python run_generation.py \
...
+### Examples
+Evaluate Llama 7B on Gaudi on task PiQA, using the BF16 data type.
+```
+python run_lm_eval.py \
+--model_name_or_path meta-llama/Llama-2-7b-hf \
+--use_hpu_graphs \
+--use_kv_cache \
+--bf16 \
+--batch_size=1 \
+--task piqa \
```
Comment (polisettyvarma) [inline examples/text-generation/README.md:243]: tasks

File: examples/text-generation/run_lm_eval.py [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+    model, tokenizer, generation_config = initialize_model(args, logger)
+
+    lm_tasks = lm_eval.tasks.get_task_dict(args.tasks)
+    lm = HabanaModelAdapter(tokenizer, model, args, generation_config)
+
+    eval_start = time.perf_counter()
+    results = lm_eval.evaluator.evaluate(lm, lm_tasks, limit=args.limit_iters)
+    import habana_frameworks.torch.hpu as torch_hpu
+
+    torch_hpu.synchronize()
```
Comment (polisettyvarma) [inline examples/text-generation/run_lm_eval.py:149]: can we do similar check like line 156

**Review by polisettyvarma — CHANGES_REQUESTED** [review]

**Review by polisettyvarma — APPROVED** [review]

**Comment by bgoldberg-habana** [comment]
@regisss can you review this also?

File: examples/text-generation/run_lm_eval.py [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+        seq_length = inps.shape[-1]
+        padding_length = 0
+        if self.options.static_shapes:
+            bucket_length = self.find_bucket(seq_length)
+            padding_length = bucket_length - seq_length
+            inps = F.pad(inps, (0, padding_length), value=self.model.config.pad_token_id)
+        logits = self.model(
+            inps.to(self._device),
+            use_cache=self.options.use_cache,
+            attn_softmax_bf16=self.options.attn_softmax_bf16,
```
Comment (regisss) [inline examples/text-generation/run_lm_eval.py:130]: This will raise an error for non-Llama models as only Llama has the `attn_softmax_bf16` arg in its forward.
It should be defined in `model.generation_config`.

**Review by regisss — COMMENTED** [review]
I left a couple of comments, only one being major as this script will fail for non-Llama models.
I also pushed a commit to fix the code blocks in the README.

File: examples/text-generation/requirements_lm_eval.txt [hunk 10]
```diff
@@ -0,0 +1 @@
+git+https://github.com/polisettyvarma/lm-evaluation-harness.git@lm_harness_fixes
```
[the remark this answers is not in the document]
Comment (bgoldberg-habana) [reply examples/text-generation/requirements_lm_eval.txt:1]: there is no intention to do that.. 
the problem they didn't release any new version for almost a year.
we hope that they release a new version with all the fixes needed. so we can use real version.

File: examples/text-generation/run_lm_eval.py (hunk 9 above)
Comment (polisettyvarma) [reply examples/text-generation/run_lm_eval.py:130]: regarding arguments to model, we can pass as per config.model_type

Comment (regisss) [reply examples/text-generation/run_lm_eval.py:130]: Yeah I agree with @polisettyvarma, in that case we should have a if-else block depending on `model.config.model_type == "llama"`. Or another solution is to inspect the signature of the forward method with `inspect.signature`: https://docs.python.org/3.8/library/inspect.html#inspect.signature

**Review by regisss — APPROVED** [review]
LGTM!
